src/GraphCoarsen.hpp

```diff
@@ -179,14 +181,63 @@ class GraphCoarsen
                         bool build_coarse_relation);
 
     /**
-       @brief take edge-based traces functions, extend them, find bubbles,
-       and assemble into interpolation matrix.
-
-       @param edge_trace is in
-       @param vertex_target is in
-       @param face_cdof is out, the face_cdof relation on coarse mesh (coarse
-       faces, coarse dofs)
-       @param Pedges is out
+       Method called from BuildPEdges
+
+       @param[in] nfaces number of faces
+       @param[in] edge_traces lives on a face
+       @param[out] face_cdof the coarseface_coarsedof relation table
+
+       @return total_num_traces on all faces
+    */
+    int BuildCoarseFaceCoarseDof(unsigned int nfaces,
+                                 std::vector<mfem::DenseMatrix>& edge_traces,
+                                 mfem::SparseMatrix& face_cdof);
+
+    /**
+       Modify the traces so that "1^T D PV_trace = 1", "1^T D other trace = 0"
+
+       Helper for BuildPEdges
+    */
+    void NormalizeTraces(std::vector<mfem::DenseMatrix>& edge_traces,
+                         const mfem::SparseMatrix& Agg_vertex,
+                         const mfem::SparseMatrix& face_edge);
+
+    /**
+       Figure out NNZ for each row of PEdges, which is to say, for each fine
+       edge dof, figure out how many coarse dofs it gets interpolated from.
+
+       @return the I array of PEdges for CSR format.
+    */
+    int* InitializePEdgesNNZ(std::vector<mfem::DenseMatrix>& edge_traces,
+                             std::vector<mfem::DenseMatrix>& vertex_target,
+                             const mfem::SparseMatrix& Agg_edge,
+                             const mfem::SparseMatrix& face_edge,
+                             const mfem::SparseMatrix& Agg_face);
+
+    /**
+       @brief takes the column column of potentials, multiplies
+       DtransferT by it, and returns the inner product with trace
+
+       helper for BuildPEdges
+    */
+    double DTTraceProduct(const mfem::SparseMatrix& DtransferT,
```

Comment:
I think this is a good name for what the function is doing. However, it may be hard to see why we want to compute this product (assuming the person only reads our paper). And apparently it is not easy to explain briefly. I will write a short note to explain this. You may see if anything we can do to make future developers/maintainers' lives easier.